# Importing Libraries


In [1]:
import pandas as pd
import numpy as np


# Data PreProcessing


In [2]:
df=pd.read_csv('diabetes.csv') 

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
df.shape

(768, 9)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [6]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

# As No Null Value Exists. PreProcessing not Required.
## But some Columns have values 0. Which is not possible.
### Lets Deal with That

In [7]:
print("Zero Values")
for col in df.columns.values:
    sum=0
    for i in range(0,len(df[col])):
        if(0==df[col][i]):
            sum+=1
    print(f"{col} :",sum)

Zero Values
Pregnancies : 111
Glucose : 5
BloodPressure : 35
SkinThickness : 227
Insulin : 374
BMI : 11
DiabetesPedigreeFunction : 0
Age : 0
Outcome : 500


# Here Glucose , BloodPressure , SkinThickness, Insulin and BMI cannot Have Value 0
## Appending them With Mean Value maybe a good

In [8]:
cols=['Glucose','BloodPressure','SkinThickness','Insulin','BMI']
for i in range(0,len(cols)):
    df[cols[i]].replace(0,df[cols[i]].median(),inplace=True)

In [9]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35     30.5  33.6   
1            1       85             66             29     30.5  26.6   
2            8      183             64             23     30.5  23.3   
3            1       89             66             23     94.0  28.1   
4            0      137             40             35    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

# Reviewing Detailed details with output

In [26]:

def Detailed_Details(df,target,outcome):
    print("Comparing",target," With",outcome)
    print("Total",len(df))
    print("\n")
    print("Column Name \t \t Total \t \t Percentage \t \t Total_Outcome(1) \t \t Percentage_Outcome(1) \t \t Total_Outcome(0)\t \tPercentage_Outcome(0) ")
    #for checking type
    type={'int','int16','int32','int64','float','float16','float32','float64'}
   
    uniq_cases=len(df)/5
    total_data=len(df)
   
    df_current=df[target]
    check_type=df_current.dtypes
    
    #for formatting
    design=0
    
    if(str(check_type) in type):
    
        if (df_current.unique().sum()<uniq_cases):
            
            uniq_sorted=sorted(df_current.unique())
            for uniq in uniq_sorted:
                
                total_outcome_1=0
                total_outcome_0=0
                for j in range(0,len(df)):
                    if((df[outcome][j]==1) and (df_current[j]==uniq)):
                        total_outcome_1+=1
                    elif((df[outcome][j]==0) and (df_current[j]==uniq)):
                        total_outcome_0+=1
                
                total=df_current.value_counts()[uniq]
                percent= round((total/total_data*100),2)
                
                percentage_outcome_1=round(((total_outcome_1/total)*100),2)
                percentage_outcome_0=round((100-percentage_outcome_1),2)
                
                if(design==0):
                    print(target)
                
                print("\t ",uniq,"\t\t",total,"\t\t",percent,"% \t\t\t ",total_outcome_1,"\t\t\t\t",percentage_outcome_1,"%\t\t\t\t",total_outcome_0,"\t\t\t\t",percentage_outcome_0,"%")
                print("\n")
                design=1
        else:
            mean_value=round(df_current.mean(),2)
            
            total_abm=0
            total_blm=0
            
            total_outcome_abm_1=0
            total_outcome_abm_0=0
            
            total_outcome_blm_1=0
            total_outcome_blm_0=0
            
            for j in range(0,len(df)):
                if(df_current[j]>=mean_value):
                    total_abm+=1
                    if(df[outcome][j]==1):
                        total_outcome_abm_1+=1
                    else:
                         total_outcome_abm_0+=1
                else:
                    total_blm+=1
                    if(df[outcome][j]==1):
                        total_outcome_blm_1+=1
                    else:
                         total_outcome_blm_0+=1
           
            percent_abm=round(((total_abm/total_data)*100),2)
            percent_blm=round((100-percent_abm),2)
            
            percent_abm_1=round(((total_outcome_abm_1/total_abm)*100),2)
            percent_abm_0=round((100-percent_abm_1),2)
            
            percent_blm_1=round(((total_outcome_blm_1/total_blm)*100),2)
            percent_blm_0=round((100-percent_blm_1),2)
            
            if(design==0):
                    print(target)
                
            print(" LessThan(",mean_value,")\t",total_blm,"\t\t",percent_blm,"% \t\t\t ",total_outcome_blm_1,"\t\t\t\t",percent_blm_1,"%\t\t\t\t",total_outcome_blm_0,"\t\t\t\t",percent_blm_0,"%")
            print(" GreaterThan(",mean_value,")\t",total_abm,"\t\t",percent_abm,"%\t\t\t\t ",total_outcome_abm_1,"\t\t\t\t",percent_abm_1,"%\t\t\t\t",total_outcome_abm_0,"\t\t\t\t",percent_abm_0,"%")
            print("\n")
            design=1
    else:
        if (df_current.nunique()<uniq_cases):
            
            for uniq in df_current.unique():
                total_outcome_1=0
                total_outcome_0=0
                for j in range(0,len(df)):
                    if((df[outcome][j]==1) and (df_current[j]==uniq)):
                        total_outcome_1+=1
                    elif((df[outcome][j]==0) and (df_current[j]==uniq)):
                        total_outcome_0+=1
                
                total=df_current.value_counts()[uniq]
                percent= round((total/total_data*100),2)
                
                percentage_outcome_1=round(((total_outcome_1/total)*100),2)
                percentage_outcome_0=round((100-percentage_outcome_1),2)
                
                if(design==0):
                    print(target)
                
                print("\t ",uniq,"\t\t",total,"\t\t",percent,"% \t\t\t ",total_outcome_1,"\t\t\t",percentage_outcome_1,"%\t\t\t",total_outcome_0,"\t\t\t",percentage_outcome_0,"%")
                print("\n")
                design=1
        else:
            print("Target value is string and have greater unique Values")
                

In [28]:
#with Pregnancies
Detailed_Details(df,'Pregnancies','Outcome')
#Diabetes seems to increase with pregnancies

Comparing Pregnancies  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
Pregnancies
	  0 		 111 		 14.45 % 			  38 				 34.23 %				 73 				 65.77 %


	  1 		 135 		 17.58 % 			  29 				 21.48 %				 106 				 78.52 %


	  2 		 103 		 13.41 % 			  19 				 18.45 %				 84 				 81.55 %


	  3 		 75 		 9.77 % 			  27 				 36.0 %				 48 				 64.0 %


	  4 		 68 		 8.85 % 			  23 				 33.82 %				 45 				 66.18 %


	  5 		 57 		 7.42 % 			  21 				 36.84 %				 36 				 63.16 %


	  6 		 50 		 6.51 % 			  16 				 32.0 %				 34 				 68.0 %


	  7 		 45 		 5.86 % 			  25 				 55.56 %				 20 				 44.44 %


	  8 		 38 		 4.95 % 			  22 				 57.89 %				 16 				 42.11 %


	  9 		 28 		 3.65 % 			  18 				 64.29 %				 10 				 35.71 %


	  10 		 24 		 3.12 % 			  10 				 41.67 %				 14 				 58.33 %


	  11 		 11 		 1.43 % 			  7 				 63.64 %				 4 				 36.36 %


	  12 		 9 		 1.17 % 			  4 				 44.44 

In [29]:
#With Glucose
Detailed_Details(df,'Glucose','Outcome')
#Greater Glucose have greater chance of having Diabetes


Comparing Glucose  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
Glucose
 LessThan( 121.66 )	 425 		 55.34 % 			  75 				 17.65 %				 350 				 82.35 %
 GreaterThan( 121.66 )	 343 		 44.66 %				  193 				 56.27 %				 150 				 43.73 %




In [30]:
#With Blood Pressure
Detailed_Details(df,'BloodPressure','Outcome')
#Greater BloodPressure has greater chance of Diabetes


Comparing BloodPressure  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
BloodPressure
 LessThan( 72.39 )	 419 		 54.56 % 			  125 				 29.83 %				 294 				 70.17 %
 GreaterThan( 72.39 )	 349 		 45.44 %				  143 				 40.97 %				 206 				 59.03 %




In [31]:
#With Skin Thickness
Detailed_Details(df,'SkinThickness','Outcome')
#Greater Skin thickness prone to greater diabetes

Comparing SkinThickness  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
SkinThickness
 LessThan( 27.33 )	 466 		 60.68 % 			  139 				 29.83 %				 327 				 70.17 %
 GreaterThan( 27.33 )	 302 		 39.32 %				  129 				 42.72 %				 173 				 57.28 %




In [32]:
#With Insulin
Detailed_Details(df,'Insulin','Outcome')
#Greater Insulin prone to greater diabetes


Comparing Insulin  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
Insulin
 LessThan( 94.65 )	 512 		 66.67 % 			  152 				 29.69 %				 360 				 70.31 %
 GreaterThan( 94.65 )	 256 		 33.33 %				  116 				 45.31 %				 140 				 54.69 %




In [33]:
#With BMI
Detailed_Details(df,'Insulin','Outcome')
#Greater BMI prone to greater diabetes


Comparing Insulin  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
Insulin
 LessThan( 94.65 )	 512 		 66.67 % 			  152 				 29.69 %				 360 				 70.31 %
 GreaterThan( 94.65 )	 256 		 33.33 %				  116 				 45.31 %				 140 				 54.69 %




In [34]:
#with DiabetesPedegreeFunction

Detailed_Details(df,'DiabetesPedigreeFunction','Outcome')
#Greater DiabetesPedegreeFunction to greater diabetes


Comparing DiabetesPedigreeFunction  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
DiabetesPedigreeFunction
 LessThan( 0.47 )	 471 		 61.33 % 			  139 				 29.51 %				 332 				 70.49 %
 GreaterThan( 0.47 )	 297 		 38.67 %				  129 				 43.43 %				 168 				 56.57 %




In [35]:
#with Age
Detailed_Details(df,'Age','Outcome')
#Greater Age to greater diabetes


Comparing Age  With Outcome
Total 768


Column Name 	 	 Total 	 	 Percentage 	 	 Total_Outcome(1) 	 	 Percentage_Outcome(1) 	 	 Total_Outcome(0)	 	Percentage_Outcome(0) 
Age
 LessThan( 33.24 )	 474 		 61.72 % 			  122 				 25.74 %				 352 				 74.26 %
 GreaterThan( 33.24 )	 294 		 38.28 %				  146 				 49.66 %				 148 				 50.34 %




# Now to Feature Selection

In [36]:
df_comp=df.loc[:,df.columns!='Outcome']
df_comp.corrwith(df['Outcome'])

Pregnancies                 0.221898
Glucose                     0.492782
BloodPressure               0.165723
SkinThickness               0.189065
Insulin                     0.148457
BMI                         0.312249
DiabetesPedigreeFunction    0.173844
Age                         0.238356
dtype: float64

# Setting certain median for correlation
    

In [37]:
print(df_comp.corrwith(df['Outcome']).median())

0.20548178630469044


# Dropping Correlation less than 0.205

In [38]:
cols=['BloodPressure','SkinThickness','Insulin','DiabetesPedigreeFunction']
for i in range(0,len(cols)):
    df.drop(cols[i],axis=1,inplace=True)

In [39]:
df.head()

Pregnancies  Glucose   BMI  Age  Outcome
0            6      148  33.6   50        1
1            1       85  26.6   31        0
2            8      183  23.3   32        1
3            1       89  28.1   21        0
4            0      137  43.1   33        1

In [40]:
df.corr()

Pregnancies   Glucose       BMI       Age   Outcome
Pregnancies     1.000000  0.128213  0.021546  0.544341  0.221898
Glucose         0.128213  1.000000  0.231400  0.266909  0.492782
BMI             0.021546  0.231400  1.000000  0.025744  0.312249
Age             0.544341  0.266909  0.025744  1.000000  0.238356
Outcome         0.221898  0.492782  0.312249  0.238356  1.000000

# Age and Pregnacies Seem to be strongly correlated with 0.544

Checking corr of each with Outcome

In [41]:
print("Pregnancies")
print(df['Pregnancies'].corr(df['Outcome']))
print("Age")
print(df['Age'].corr(df['Outcome']))

Pregnancies
0.2218981530339867
Age
0.23835598302719752


# As Age is more correlated with Outcome. Dropping Pregnancies

In [42]:
df.drop('Pregnancies',axis=1,inplace=True)

In [43]:
df.corr()

Glucose       BMI       Age   Outcome
Glucose  1.000000  0.231400  0.266909  0.492782
BMI      0.231400  1.000000  0.025744  0.312249
Age      0.266909  0.025744  1.000000  0.238356
Outcome  0.492782  0.312249  0.238356  1.000000

In [44]:
df.head()


Glucose   BMI  Age  Outcome
0      148  33.6   50        1
1       85  26.6   31        0
2      183  23.3   32        1
3       89  28.1   21        0
4      137  43.1   33        1

# Now building Model

In [45]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.model_selection import GridSearchCV

# Now  Implementing Machine Learning


In [46]:
svc=SVC()
knn=KNeighborsClassifier()
dtc=DecisionTreeClassifier()
nb=GaussianNB()

In [47]:
models={'svc':svc,'knn':knn,'dtc':dtc,'nb':nb}

In [48]:
x,y=df.drop('Outcome',axis=1),df['Outcome']

In [49]:
train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=200)

In [50]:
model_accuracy={}

for model in models.keys():
    mod = models[model].fit(train_x,train_y)
    predict=mod.predict(test_x)
    #print(f"{model}",classification_report(predict,test_y))
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

svc
[[98 21]
 [ 8 27]]
knn
[[87 18]
 [19 30]]
dtc
[[76 25]
 [30 23]]
nb
[[96 19]
 [10 29]]


In [56]:
model_accuracy
#Same Accurracy without dropping any features

{'svc': 0.8116883116883117,
 'knn': 0.7727272727272727,
 'dtc': 0.7597402597402597,
 'nb': 0.8116883116883117}

# Lets try With GridSearch


In [52]:
params={
    'knn': {'n_neighbors':[3,5,7,9,11,13,15],
    'metric':['cosine','euclidean','manhattan'],
    'weights':['uniform','distance']},
    
    'svc': {
    'C':[0.1,1,10,100], 
    'gamma':[1,0.1,0.01,0.01], 
    'kernel':['rbf','linear']},
    
    'dtc':{
        'criterion':['gini','entropy'],
        'max_depth':[2,4,6,8,10,12]
    },
    
    'nb':{
        'priors': [None],
    'var_smoothing': [0.00000001, 0.000000001, 0.00000001]
    }
    
    }

In [53]:
model_accuracy={}
for model in models.keys():
    mod = GridSearchCV(
    models[model],
    params[model],
    verbose=1, #Progress bar showing
    cv=20, #cross validation
    n_jobs=-1, #cores to assign
)
    gridsearch_result=mod.fit(train_x,train_y)
    predict=mod.predict(test_x)
    print(f"{model} : ",gridsearch_result.best_estimator_)
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

Fitting 20 folds for each of 32 candidates, totalling 640 fits
svc :  SVC(C=10, gamma=1, kernel='linear')
svc
[[96 19]
 [10 29]]
Fitting 20 folds for each of 42 candidates, totalling 840 fits
knn :  KNeighborsClassifier(metric='euclidean', n_neighbors=15)
knn
[[92 21]
 [14 27]]
Fitting 20 folds for each of 12 candidates, totalling 240 fits
dtc :  DecisionTreeClassifier(criterion='entropy', max_depth=4)
dtc
[[77  8]
 [29 40]]
Fitting 20 folds for each of 3 candidates, totalling 60 fits
nb :  GaussianNB(var_smoothing=1e-08)
nb
[[96 19]
 [10 29]]


In [55]:
model_accuracy
#Though hyperplanning was fast after feature Selection

{'svc': 0.8116883116883117,
 'knn': 0.7727272727272727,
 'dtc': 0.7597402597402597,
 'nb': 0.8116883116883117}